File: py_drill/Problems/TR01_ValidateBST.py
# REF : https://leetcode.com/problems/validate-binary-search-tree/
# NOTES :
#       Need to compare given node with updated UPPER BOUND and LOWER BOUND
#       Like "Count good nodes in BT" deque stack will have now two more elements
#       [node, low, high] to implement iterative DFS/BFS solution

from collections import deque

# T: O(n)
# S: O(n)
# DFS recursive
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isValidBST(self, root) -> bool:

        def valid(node, low, high):
            if node is None:
                return True

            if not (node.val < high and node.val > low):
                return False

            return (valid(node.left, low, node.val) and
                    valid(node.right, node.val, high))

        return valid(root, float("-inf"), float("inf"))


# T: O(n)
# S: O(n)
# DFS iterative & BFS iterative with popleft change
class Solution2:
    def isValidBST(self, root) -> bool:
        if root is None:
            return True

        s = deque()
        s.append([root, float("-inf"), float("inf")])
        while s:
            node, low, high = s.pop()                       # BFS with popleft
            if not (node.val < high and node.val > low):
                return False
            if node.left:
                s.append([node.left, low, node.val])
            if node.right:
                s.append([node.right, node.val, high])

        return True


def main():
    pass

if __name__ == "__main__":
    main()

File: cpp_drill/Problems/TR01_ValidateBST.cpp
#include <iostream>
#include <limits>
using namespace std;

//  REF : https://leetcode.com/problems/validate-binary-search-tree/
//  NOTES :
//        Need to compare given node with updated UPPER BOUND and LOWER BOUND
//        Like "Count good nodes in BT" deque stack will have now two more elements
//        [node, low, high] to implement iterative DFS/BFS solution
//  O(T) : O(n)
//  O(S) : O(n)

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
  };

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        // INT_MIN and INT_MAX is not working for all testcases
        return valid(root, INT_MIN, INT_MAX);
    }

    bool valid(TreeNode* node, int low, int high) {
        if (node == nullptr)
            return true;

        if !(node->val > low && node->val < high)
            return false;

        // This can also work
        // if (node->val <= low || node->val >= high)
        //     return false;

        return valid(node->left, low, node->val) && valid(node->right, node->val, high);
    }
};

class Solution2 {
public:
    bool isValidBST(TreeNode* root) {
        return valid(root, nullptr, nullptr);
    }

    bool valid(TreeNode* node, TreeNode* low, TreeNode* high) {
        if (node == nullptr)
            return true;
        
        if ((low != nullptr && node->val <= low->val) ||
            (high != nullptr && node->val >= high->val))
            return false;

        return valid(node->left, low, node) && valid(node->right, node, high);
    }
};

class Solution3 {
public:
    bool isValidBST(TreeNode* root) {
        long long maxi = LONG_MAX;      // Only this is the difference.
        long long mini = LONG_MIN;
        return valid(root, mini, maxi);
    }

    bool valid(TreeNode* node, long long low, long long high) {
        if (node == nullptr)
            return true;

        if (!(node->val > low && node->val < high))
            return false;
        // if (node->val <= low || node->val >= high)
        //     return false;

        return valid(node->left, low, node->val) && valid(node->right, node->val, high);
    }
};